text/0000-import-trait-methods.md

```diff
@@ -0,0 +1,199 @@
+- Feature Name: `import-trait-methods`
+- Start Date: 2024-03-19
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Allow importing methods from traits and then using them like regular functions.
+
+# Motivation
+[motivation]: #motivation
+
+There has for a long time been a desire to shorten the duplication needed to access certain methods, such as `Default::default`. Codebases like [Bevy](https://github.com/bevyengine/bevy/blob/7c7d1e8a6442a4258896b6c605beb1bf50399396/crates/bevy_utils/src/default.rs#L27) provide wrapper methods to shorten this call, and a previous, now-rejected, [RFC](https://github.com/rust-lang/rust/pull/73001) aimed to provide this method as part of the standard library. This RFC was rejected with a note that there is a desire for a more general capability to import any trait method.
+
+Additionally, if you pull in a crate like [num_traits](https://docs.rs/num-traits/latest/num_traits/), then this feature will allow access to numeric methods such as `sin` using the infix syntax that is more common in mathematics. More generally, it will make infix calls to a trait method shorter without having to write a wrapper function.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Importing a method from a trait is the same as importing a method from any module:
+```rust
+use Default::default;
+```
+
+Once you have done this, the method is made available in your current scope just like any other regular function.
+
+```rust
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: default()
+        }
+    }
+}
+```
+
+You can also use this with trait methods that take a `self` argument:
+
+```rust
+use num_traits::float::Float::{sin, cos}
+
+fn eulers_formula(theta: f64) -> (f64, f64) {
+    (cos(theta), sin(theta))
+}
+```
+
+Importing a method from a trait does not import the trait. If you want to call `self` methods on a trait or `impl` it, then you can import the trait as well as methods in the trait:
+
+```rust
+mod a {
+    trait A {
+        fn new() -> Self;
+        fn do_something(&self);
+    }
+}
+
+mod b {
+    use super::a::A::{self, new}
+
+    struct B();
+
+    impl A for B {
+        fn new() -> Self {
+            B()
+        }
+
+        fn do_something(&self) {
+        }
+    }
+
+    fn f() {
+        let b: B = new();
+        b.do_something();
+    }
+}
+```
+
+You cannot import a parent trait method from a sub-trait.
```

Comment:
I think we should allow importing a parent trait method from a sub-trait, since it's not uncommon to have a parent trait that isn't ever intended to be used except through its sub-trait, this is used as a method to guarantee that the parent trait methods always have the default implementation rather than being overridable, e.g.:
```rust
mod private {
    use super::*;
    pub trait Parent {
        fn as_subtrait(&self) -> &dyn Subtrait;
    }
}

impl<T: Subtrait> private::Parent for T {
    fn as_subtrait(&self) -> &dyn Subtrait {
        self
    }
}

pub trait Subtrait: Parent {
}
```

Reply:
Is your pattern intended to only work intra-crate? I can't find any way to call `as_subtrait` in your example from another crate: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=637ea2991a5a3c4e075cbaa365b643e8

Reply:
hmm, i was under the impression it would work if `Parent` is `pub` (which it isn't in your example)...

Reply:
Minimized it slightly better: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f5535ec3980f6c4b60e62c86d1437071. Here everything but `private` is `pub` and it doesn't work.